hybrid/compartment_examples/inter_comp_call_secure/main.c

```diff
@@ -0,0 +1,194 @@
+/* In this example, we tighten security even further. For each compartment we
+ * create, we store some capabilities of interest. Essentially, we setup
+ * certain entry points, and make them available to each compartment.
+ * Currently, these involve the PCC/DDC of the `switch_compartment` function
+ * (PCC to execute, DDC to access memory where we store information pertaining
+ * to compartments; essentially, `switch_compartment` performs a limited
+ * privelege escalation, but only over prepared data within `main`).
+ * Additionally, we also seal the PCC of `switch_compartment`.
+ *
+ * Note that in the current design, each compartment has a single function it
+ * executes when it is called. Allowing multiple entry points can be modelled
+ * by having more capabilities be saved, in addition to the compartment
+ * switching ones, in each compartment's heap. Alternatively, we could, within
+ * the compartment switcher, determine in which compartment we need to switch
+ * to call a given function, based on PCC bounds for each compartment.
+
+ * Additionally, we also changed the expected memory layout in the
+ * compartments. In the absence of a custom memory allocator, we note that the
+ * heap is not used, except for manually storing capabilities as detailed
+ * above. Current compartment memory layout:
+ *
+ * `DDC value` > -----------------
+ *              |                 |
+ *              |     HEAP        |
+ *              |                 |
+ *              |   saved caps    |
+ *         sp > |-----------------| < DDC + total_comp_size - comp_stack_size
+ *              |                 |
+ *              |     STACK       |
+ *               -----------------  < DDC + total_comp_size
+ */
+
+#include "../../../include/common.h"
+#include "../../include/utils.h"
+
+#include <assert.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#if !defined(__CHERI_CAPABILITY_WIDTH__) || defined(__CHERI_PURE_CAPABILITY__)
+#error "This example only works on CHERI hybrid mode"
+#endif
+
+/*******************************************************************************
+ * Assembly Functions & Consts
+ ******************************************************************************/
+
+extern int switch_compartment();
+extern void comp_f_fn();
+extern void comp_g_fn();
+
+/*******************************************************************************
+ * Types & Consts
+ ******************************************************************************/
+
+const size_t comp_stack_size = 1992;
+const size_t total_comp_size = 5000;
+size_t id = 0;
+
+// Duplicate certain capabilities within the heap of each compartments;
+// currently we save DDC and PCC of `switch_compartment`
```

Comment:
If another compartment can arbitrarily read/write to the compartment switcher I think, in general, that it can get the compartment switcher to do whatever it wants. But, if `switch_compartment` is a sentry (which I *think* it is), we don't need its PCC to be stored outside? The DDC is an interesting question: I'm not sure how to handle that. Can a sentry also change the DDC?

Reply:
Sorry, I don't quite follow. I think there are three issues at play here:
1. The minimal amount of data a malicious compartment requires in order to be able to "take control" of `switch_compartment` (sentence 1).
2. Requiring the PCC of `switch_compartment` to be "stored outside" (I don't understand where "outside" is). (sentence 2)
3. Issue of DDC integrity (sentence 3-4)

For 1, we need to balance the minimal information `switch_compartment` needs in order to actually perform the switch. We obviously need *something* to be exposed in order for the function to be executable (at minimum, we can say the address of `switch_compartment`, but of course that is useless without knowing what compartment we want to switch to, or where that compartment is in memory). If the minimal information needed to perform the switch and the minimal data needed to take control of `switch_compartment` are not disjoint, then this is a design problem, and this design is infeasible.

For 2, I will clarify that the `switch_compartment` available to the compartments is a sentry, which can only be branched to. I need further clarification regarding the "outside" thing. For 3, what does "a sentry [..] change the DDC" mean? A sentry is a sealed capability with it's value being the address of a function. By that definition, a sentry is changing the DDC, as the DDC swapping happens inside `switch_compartment` (although the DDC is expected in a register, provided as an argument by the calling compartment).

Reply:
> although the DDC is expected in a register, provided as an argument by the calling compartment

I think that's a concern. What stops the caller from providing an inappropriate DDC? However, in this case I think it can only cause a crash in the switcher (by providing an inadequate DDC). That might be acceptable in the (currently implicit) threat model.

If it's necessary to bind a pair (to prevent this), it's possible to do that using `ldpblr`, as we've discussed before.